File: CalSnap/Features/MealScanner/EditableFoodItem.swift
import Foundation

struct EditableFoodItem: Identifiable, Equatable {
    let id: UUID
    var name: String
    var weightG: Double
    var calories: Int
    var proteinG: Double
    var carbsG: Double
    var fatG: Double
    var fiberG: Double
    var confidence: Double
    var isFlagged: Bool
    var originalWeightG: Double?

    init(
        id: UUID = UUID(),
        name: String,
        weightG: Double,
        calories: Int,
        proteinG: Double,
        carbsG: Double,
        fatG: Double,
        fiberG: Double,
        confidence: Double,
        isFlagged: Bool,
        originalWeightG: Double? = nil
    ) {
        self.id = id
        self.name = name
        self.weightG = weightG
        self.calories = calories
        self.proteinG = proteinG
        self.carbsG = carbsG
        self.fatG = fatG
        self.fiberG = fiberG
        self.confidence = confidence
        self.isFlagged = isFlagged
        self.originalWeightG = originalWeightG ?? weightG
    }

    mutating func updateWeight(to newWeightG: Double) {
        guard weightG > 0 else { return }
        let ratio = newWeightG / weightG
        calories = Int((Double(calories) * ratio).rounded())
        proteinG *= ratio
        carbsG *= ratio
        fatG *= ratio
        fiberG *= ratio
        weightG = newWeightG
    }

    static func from(
        result: MealAnalysisResponse.FoodItemResult,
        flaggedNames: Set<String>
    ) -> EditableFoodItem {
        let isFlagged = result.confidence < AppConstants.Gemini.confidenceThreshold
            || flaggedNames.contains(result.name)
        return EditableFoodItem(
            name: result.name,
            weightG: result.estimatedWeightG,
            calories: result.calories,
            proteinG: result.proteinG,
            carbsG: result.carbsG,
            fatG: result.fatG,
            fiberG: result.fiberG,
            confidence: result.confidence,
            isFlagged: isFlagged,
            originalWeightG: result.estimatedWeightG
        )
    }

    static func emptyManual() -> EditableFoodItem {
        EditableFoodItem(
            name: "",
            weightG: 100,
            calories: 0,
            proteinG: 0,
            carbsG: 0,
            fatG: 0,
            fiberG: 0,
            confidence: 1.0,
            isFlagged: false,
            originalWeightG: 100
        )
    }

    func toFoodItem() -> FoodItem {
        FoodItem(
            id: id,
            name: name,
            estimatedWeightG: weightG,
            calories: calories,
            proteinG: proteinG,
            carbsG: carbsG,
            fatG: fatG,
            fiberG: fiberG,
            confidence: confidence,
            usdaFoodId: nil,
            isFlagged: isFlagged
        )
    }
}

File: calsnap-web/lib/scanner/editable-food-item.ts
import { AppConstants } from '@/lib/constants';
import type { FoodItem } from '@/lib/models/food-item';
import type { MealAnalysisFoodItemResult } from '@/lib/gemini/meal-analysis-types';

export interface EditableFoodItem {
  id: string;
  name: string;
  weightG: number;
  calories: number;
  proteinG: number;
  carbsG: number;
  fatG: number;
  saturatedFatG: number;
  unsaturatedFatG: number;
  fiberG: number;
  confidence: number;
  isFlagged: boolean;
  originalWeightG: number;
}

export function updateEditableItemWeight(
  item: EditableFoodItem,
  newWeightG: number,
): EditableFoodItem {
  if (item.weightG <= 0) {
    return item;
  }
  const ratio = newWeightG / item.weightG;
  return {
    ...item,
    weightG: newWeightG,
    calories: Math.round(item.calories * ratio),
    proteinG: item.proteinG * ratio,
    carbsG: item.carbsG * ratio,
    fatG: item.fatG * ratio,
    saturatedFatG: item.saturatedFatG * ratio,
    unsaturatedFatG: item.unsaturatedFatG * ratio,
    fiberG: item.fiberG * ratio,
  };
}

export function editableFoodItemFromFoodItem(foodItem: FoodItem): EditableFoodItem {
  return {
    id: foodItem.id,
    name: foodItem.name,
    weightG: foodItem.estimatedWeightG,
    calories: foodItem.calories,
    proteinG: foodItem.proteinG,
    carbsG: foodItem.carbsG,
    fatG: foodItem.fatG,
    saturatedFatG: foodItem.saturatedFatG,
    unsaturatedFatG: foodItem.unsaturatedFatG,
    fiberG: foodItem.fiberG,
    confidence: foodItem.confidence,
    isFlagged: foodItem.isFlagged,
    originalWeightG: foodItem.estimatedWeightG,
  };
}

export function editableFoodItemFromAnalysisResult(
  result: MealAnalysisFoodItemResult,
  flaggedNames: Set<string>,
): EditableFoodItem {
  const isFlagged =
    result.confidence < AppConstants.Gemini.confidenceThreshold ||
    flaggedNames.has(result.name);
  return {
    id: crypto.randomUUID(),
    name: result.name,
    weightG: result.estimatedWeightG,
    calories: result.calories,
    proteinG: result.proteinG,
    carbsG: result.carbsG,
    fatG: result.fatG,
    saturatedFatG: result.saturatedFatG,
    unsaturatedFatG: result.unsaturatedFatG,
    fiberG: result.fiberG,
    confidence: result.confidence,
    isFlagged,
    originalWeightG: result.estimatedWeightG,
  };
}

export function itemWeightRange(originalWeightG: number): { min: number; max: number } {
  const min = Math.max(1, Math.round(originalWeightG * 0.5));
  const max = Math.max(min + 1, Math.round(originalWeightG * 1.5));
  return { min, max };
}

export function editableFoodItemToFoodItem(item: EditableFoodItem): FoodItem {
  return {
    id: item.id,
    name: item.name,
    estimatedWeightG: item.weightG,
    calories: item.calories,
    proteinG: item.proteinG,
    carbsG: item.carbsG,
    fatG: item.fatG,
    saturatedFatG: item.saturatedFatG,
    unsaturatedFatG: item.unsaturatedFatG,
    fiberG: item.fiberG,
    confidence: item.confidence,
    usdaFoodId: undefined,
    isFlagged: item.isFlagged,
  };
}
